### ISSUE TITLE
Claims bus name before device are fully enumerated

### ISSUE THREAD

**Issue by wjt** [storaged-project/udisks#635]
I have a service which, on startup, looks for a `/dev/mmcblk*boot0` device using UDisks, and changes its behaviour based on whether or not it can be found. My code previously assumed that when `udisks_client_new()`'s callback is called, and assuming `udisks_client_new_finish()` succeeds, the resulting `UDisksClient` object's ObjectManager knows about all block devices in the system.

This turns out not to be true: sometimes, the device is announced (by InterfacesChanged) after UDisks has appeared on the bus, and hence after `udisks_client_new()` has called its callback. I had a dig through the code. udisks_provider_start() is called from udisks_daemon_constructed() which is called from on_bus_acquired() (which is before the bus name is claimed). It appears that it does try to enumerate all block devices known to libgudev:

https://github.com/storaged-project/udisks/blob/71cd4c6b674e8e9458498afb2f82df6e316c04bc/src/udiskslinuxprovider.c#L679-L690

although devices which are not "initialized" are skipped:

https://github.com/storaged-project/udisks/blob/71cd4c6b674e8e9458498afb2f82df6e316c04bc/src/udiskslinuxprovider.c#L464

Running `udevadm monitor -s block` from before the point where UDisks is started, I see that this event (and similar events for other partitions on the same device) shortly after UDisks starts:

Feb 20 14:54:33 ecs udevadm[443]: UDEV  [14.898602] add      /devices/pci0000:00/80860F14:00/mmc_host/mmc0/mmc0:0001/block/mmcblk0/mmcblk0boot0 (block)

and right after that, of course, UDisks reports the existence of that device.

I can work around this in my service by waiting for UDisks to notice the relevant devices (if my application determines that they're present, using `glob()` – I can't just open them directly because the service does not run as root) but it would be nice if UDisks could do more to ensure existing devices are fully enumerated before claiming its bus name. I'm not sure how best to accomplish this, though.

**Comment by tbzatek** [comment]
I see your point. Generally `udisks` enumerates udev database on startup but can't afford to wait until all devices are initialized. Probing may take significant amount of time for each device. Please note that some `udisks` consumers may possibly wait for the service to come up in a busy wait so this operation should be as fast as possible. The current `udisks` startup code is still a subject to further optimization.

In your case when the block device is known to udev but not yet fully probed, would calling `udevadm settle` fix your problem? (apparently this can be called unprivileged). Note that even after that this is still racy as `udisks` performs its own probing and d-bus object export may take some time too.

In the worst case we may think of providing some kind of similar convenient method call that would block until all known block devices are properly initialized and exported on D-Bus (taking an optional timeout and exporting a cancellable job).

@vojtechtrefny, @vpodzime: thoughts?

**Comment by wjt** [comment]
Sure, I understand that this is an unusual case and that most apps or services using UDisks are happy to detect new devices as they appear. If I'd realised sooner that this race would happen, I might have designed the way the service initializes itself a bit differently, but I'm a fair way down the wrong path at this point.

> In your case when the block device is known to udev but not yet fully probed, would calling `udevadm settle` fix your problem? (apparently this can be called unprivileged). Note that even after that this is still racy as `udisks` performs its own probing and d-bus object export may take some time too.

That was my initial workaround for this problem, but we decided against it because:

1. we only need to wait for block devices to be enumerated, not for all devices on the system to be fully probed
2. that potential race was pointed out to me!

> In the worst case we may think of providing some kind of similar convenient method call that would block until all known block devices are properly initialized and exported on D-Bus (taking an optional timeout and exporting a cancellable job).

Sounds perfect for this application. (I briefly hoped that this is what `udisks_client_settle` did.)

**Comment by vpodzime** [comment]
> In the worst case we may think of providing some kind of similar convenient method call that would block until all known block devices are properly initialized and exported on D-Bus (taking an optional timeout and exporting a cancellable job).

I think that makes sense. And even for some other cases, potentially, like adding many iSCSI devices and similar.

**Comment by tbzatek** [comment]
Right, the `udisks_client_settle()` call is little different and perhaps misleading in name, it stands more for _settle the client_, not _the daemon_. Function naming is important indeed.
